feat(iota-genesis-builder,iota-node): Genesis restructuring

miker83z opened the pull request:
# Description of change

This PR implements a data migration mechanism that modifies the way the genesis is built and the way the node is started for the first time.
It fixes two main issues described in #2878:
- It avoids the creation of a huge transaction that can be fetched via RPC and DDOS nodes.
- it reduces the RAM usage on nodes during startup.

The changes are:
- `genesis.blob` now doesn't contain object coming from the Stardust migration.
- It contains a checkpoint that includes a list of transaction digests and transaction effects digests; among these, 1 transaction is the "main" genesis transaction that creates system packagess and token allocations, whilst, 9999 transactions are there to push Stardust migration objects into the state.
- These so called 9999 "migration transactions" are genereted during the genesis-build process and stored in a separate file `migration.blob`.
- During all startup operations, the node only uses the data contained in the `genesis.blob`; only when the `AuthorityStore` DB is opened for the first time, `migration.blob` is read and its transactions/objects pushed into the DB (the node uses `migration.blob` in other 2 not so relevant occasions, in the same startup context). 

## Links to any relevant issues

Closes #2878 

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## How the change has been tested

1. `cargo run --release --bin iota genesis --working-dir ../test_iota_config -f --with-faucet --num-validators 1 --remote-migration-snapshots https://stardust-objects.s3.eu-central-1.amazonaws.com/iota/alphanet/latest/stardust_object_snapshot.bin.gz` 
(for using a local snapshot `cargo run --release --bin iota genesis --working-dir ../test_iota_config -f --with-faucet --num-validators 1 --local-migration-snapshots ../stardust_object_snapshot.bin`)
3. `cargo run --release --bin iota start --network.config ../test_iota_config`

## Change checklist

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes

Dkwcs committed:
feat(iota-genesis-builder, iota-core): support migration transactions effects in genesis (#2895)

* feat(iota-genesis-builder): add dedicated function to split and burn migrated objects (#2885)
* feat(iota-config, iota-genesis-builder): accumulate migration txs effect in genesis in order to verify migration blob data in future
* feat(iota-genesis-builder): objects chunk per transaction implementation

Co-authored-by: Mirko Zichichi <miker83z@proton.me>

Dkwcs committed:
feat(iota-genesis-builder, iota-swarm, iota-core, iota-cluster): generating of the migration.blob for the migration txs data accumulation (#2925)

* feat(iota-genesis-builder, iota-swarm, iota-core, iota-cluster): generating of the migration.blob for the migration txs data accumulation

kodemartin committed:
fix(iota-genesis-builder): restore save/load operations of migration tx data (#3028)

* fix(iota-genesis-builder): fix save/load of migration tx data in the builder

* refactor(iota-genesis-builder): pass migration sources from cli in from_s3 example

Dkwcs committed:
feat(iota-genesis-builder, iota-swarm, iota-core, iota-config): Store migration data in the node's database (#2955)

* feat(iota-genesis-builder): add dedicated function to split and burn migrated objects (#2885)

* fix(iota-config, iota-genesis-builder): accumulate migration txs effect in genesis in order to verify migration blob data in future

* fix(iota-genesis-builder): objects chunk per transaction implementation

* Removing extra println

* fix(iota-genesis-builder, iota-swarm, iota-core, iota-cluster): generating of the migration.blob for the migration txs data accumulation

* fix(iota): Deal with migration tx data for network config (#2937)

* fix(iota): remove migration tx data from network config

* fix(iota): save migration tx data

* fix(cluster-test): comment

* fix(iota-genesis-builder, iota-swarm, iota-core, iota-config): store migration data in the node's db, pass config_dir field in order to get acess to the migration data

* fix(iota-core): add error message to migration data storing process

* Fix review comments

* fix review comments

* Fix review comments

* refactor(node-config): optimizations and clippy

* fix(iota-config, iota-genesis-builder): fix issue #2957(genesis.blob to only contain TransactionDigests)

* feat(iota-config/node): create a node::MigrationTxData type (#3023)

* feat(iota-config/node): create a node::MigrationTxData type

* feat(iota-config/node): make node::MigrationTxData memory efficient

* fix(node): serde serialization

* fix(node): pass migration objects to AuthorityState (#3030)

* feat(iota-genesis-builder): remove migration tx digests from Genesis (#3034)

* fix(iota-node): migration transactions execution has been added (#3032)

* fix(iota-node): migration transactions execution have been added
---------
Co-authored-by: miker83z <mirko.zichichi@iota.org>

* fix(iota-genesis-builder): remove leftover

---------

Co-authored-by: Mirko Zichichi <miker83z@proton.me>
Co-authored-by: miker83z <mirko.zichichi@iota.org>

kodemartin reviewed (changes requested):
Hey @miker83z @Dkwcs great job overall!

Went through `iota_genesis_builder` as a first review iteration, and left a few suggestions. Most are minor, expect for the one about validating `MigrationTxData` while building `Genesis`. This relates to the WIP in #3041.

Will go over the rest of the implementation once #3041 is finished.

Dkwcs committed:
feat(iota-node, iota-config, iota-core, genesis-builder): remove vector of objects from the migration blob (#3052)

* feat(iota-node, iota-config, iota-core, genesis-builder): remove vector of objects from the migration blob, get objects on the fly
* feat(iota-core, iota-config): migration.blob content validation

Dkwcs committed:
Fix review comments

kodemartin committed:
fix: resolve conflicts with rebase

kodemartin committed:
feat: update restructured genesis to mainnet v1.32.2 (#3135)

kodemartin committed:
Revert "feat: update restructured genesis to mainnet v1.32.2 (#3135)"

This reverts commit e6e5cdcf1d7dd703c39bf58040c9ab12d7e5f79d.

Dkwcs committed:
Fix review comments

kodemartin committed:
Merge pull request #3138 from iotaledger/sc-platform/restructured-genesis-fix-conflicts

feat: update restructured genesis to mainnet v1.32.2

Dkwcs committed:
Extra method has been removed

Dkwcs committed:
fix(iota-genesis-builder): Refactoring chunk processing for migration objects

Dkwcs committed:
cargo fmt

Dkwcs committed:
Minor fix

kodemartin reviewed (approved):
Looks good to me! Left a few final refactoring suggestions that I hope you find plausible.

Dkwcs committed:
Update crates/iota-node/src/lib.rs

Co-authored-by: Konstantinos Demartinos <konstantinos.demartinos@iota.org>

Dkwcs committed:
Review comments fixes:
1. error/logging message begin with lower-case
2. New method for migration_tx_data (get_objects) getting all objects
3. Minor authority db migration tx loops refactoring

Dkwcs committed:
Review comments fixes

Dkwcs committed:
Fix review comments

Dkwcs committed:
Invert contains_migrations logic

Dkwcs committed:
Renaming "vanilla" to "without_migration" suffix

Dkwcs committed:
cargo fmt

Dkwcs committed:
Update crates/iota-genesis-builder/src/lib.rs

Co-authored-by: Mirko Zichichi <miker83z@proton.me>

Dkwcs committed:
Update crates/iota-genesis-builder/src/lib.rs

Co-authored-by: Mirko Zichichi <miker83z@proton.me>

Dkwcs committed:
Update crates/iota-config/src/genesis.rs

Co-authored-by: Mirko Zichichi <miker83z@proton.me>

lzpap committed:
fix(iota-node): lazy load migration.blob file in node (#3165)

* fix(iota-node): lazy load migration.blob file in node

* Propegate error instead of `expect()`

Co-authored-by: Mirko Zichichi <miker83z@proton.me>

---------

Co-authored-by: Mirko Zichichi <miker83z@proton.me>

miker83z committed:
feat(genesis-builder): validate migration data before taking it (#3162)

* refactor(iota-genesis-builder): build method returns new type GenesisBuildEffects that accumulates both genesis and migration data

miker83z committed:
Merge branch 'develop' into sc-platform/reduce_migrated_genesis_size

Dkwcs committed:
Review comments fix

Dkwcs committed:
Cargo fmt

lzpap committed:
fix(iota-genesis-builder): fix examples

Dkwcs committed:
Fix of review comments

Dkwcs committed:
Fix of review comments

Dkwcs committed:
remove extra serde arg

Dkwcs committed:
fix(iota-config): tests fix (#3217)

* fix(iota-config): iota-config  tests fix

Dkwcs committed:
fix(iota-swarm-config): update config snapshot with mock migration blob path (#3222)

miker83z committed:
Merge branch 'develop' into sc-platform/reduce_migrated_genesis_size

miker83z committed:
fix/refactor(iota-genesis-builder): Create `iota-genesis-common` and fix migration objects versioning (#3228)

* refactor(iota-genesis-builder): create iota-genesis-common

* fix(iota-config): update migration Move objects versions

* fix(iota-genesis-builder): after rebase

* fix(iota-genesis-common): fix dprint

* docs(iota-genesis-common): add comments

* feat(iota-config): validate migration objects execution

* fix(iota-genesis-common): remove unused dependency

---------

Co-authored-by: Levente Pap <levente.pap@iota.org>

miker83z committed:
Merge branch 'develop' into sc-platform/reduce_migrated_genesis_size

miker83z committed:
docs(iota-genesis): add comments

miker83z committed:
refactor(iota-genesis-builder): migration_tx_max_amount

miker83z committed:
refactor(iota-config): remove node's version of MigrationTxData

Dkwcs committed:
Update crates/iota-genesis-builder/src/lib.rs

Co-authored-by: Thoralf-M <46689931+Thoralf-M@users.noreply.github.com>

miker83z committed:
docs(iota-genesis): fix comments

miker83z committed:
Merge branch 'develop' into sc-platform/reduce_migrated_genesis_size

Dkwcs committed:
fix(iota-swarm-config): snapshot update

miker83z committed:
fix(ci): add iota-genesis-common to crates-filters

miker83z committed:
refactor(iota-genesis-builder): remove SMR

Co-authored-by: Thoralf-M <46689931+Thoralf-M@users.noreply.github.com>

miker83z committed:
Merge branch 'develop' into sc-platform/reduce_migrated_genesis_size

miker83z committed:
Merge branch 'develop' into sc-platform/reduce_migrated_genesis_size

Dkwcs committed:
Update crates/iota-config/src/genesis.rs

Co-authored-by: DaughterOfMars <chloedaughterofmars@gmail.com>

Dkwcs committed:
Update crates/iota-genesis-common/src/lib.rs

Co-authored-by: Thoralf-M <46689931+Thoralf-M@users.noreply.github.com>

miker83z committed:
Merge branch 'develop' into sc-platform/reduce_migrated_genesis_size

miker83z committed:
Update crates/iota-genesis-builder/src/lib.rs

Co-authored-by: Thoralf-M <46689931+Thoralf-M@users.noreply.github.com>

miker83z committed:
comments

miker83z committed:
fix copyright

miker83z committed:
fix the 'l' key on my keyboard

miker83z committed:
fix dependencies and full node template

miker83z committed:
fix email contact

miker83z committed:
Merge branch 'develop' into sc-platform/reduce_migrated_genesis_size
